### PR TITLE
feat: issue griefing collateral in selectable currency

### PR DESCRIPTION
**Pull request by sander2**
To accompany https://github.com/interlay/interbtc/pull/1054 

I tested the migration with chopsticks.

### CODE REVIEW

File: crates/issue/src/lib.rs [hunk 1]
```diff
@@ -316,11 +321,8 @@ impl<T: Config> Pallet<T> {
         ext::vault_registry::ensure_not_banned::<T>(&vault_id)?;
 
         // calculate griefing collateral based on the total amount of tokens to be issued
-        let amount_collateral = amount_requested.convert_to(T::GetGriefingCollateralCurrencyId::get())?;
```
Comment (daniel-savu) [inline crates/issue/src/lib.rs:319]: Shouldn't you also remove the `GetGriefingCollateralCurrencyId` associated type from the vault-registry? It's supposed to not be used at all any longer, and that will help point out if other parts of the code should also be changed

Comment (daniel-savu) [reply crates/issue/src/lib.rs:319]: I noticed that `nomination`, `redeem` and `replace` still use it and it's no longer present in the `issue` pallet, so I guess it's fine. Would be cleaner to have the same griefing collateral everywhere though, but that's a product decision.

File: crates/issue/src/migration.rs [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+use sp_core::H256;
+
+/// The log target.
+const TARGET: &'static str = "runtime::issue::migration::v1";
+
+/// The original data layout of the democracy pallet without a specific version number.
+mod v0 {
+    use super::*;
+
+    // Due to a known bug in serde we need to specify how u128 is (de)serialized.
```
Comment (daniel-savu) [inline crates/issue/src/migration.rs:12]: Does this comment still apply? I can't find this deserialization configured anywhere

File: crates/issue/src/migration.rs [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+
+    pub struct Migration<T>(sp_std::marker::PhantomData<T>);
+
+    impl<T: Config + frame_system::Config<Hash = H256>> OnRuntimeUpgrade for Migration<T> {
+        #[cfg(feature = "try-runtime")]
+        fn pre_upgrade() -> Result<Vec<u8>, &'static str> {
+            assert_eq!(StorageVersion::get::<Pallet<T>>(), 0, "can only upgrade from version 0");
+
+            let issue_count = v0::IssueRequests::<T>::get().len();
+            log::info!(target: TARGET, "{} public proposals will be migrated.", issue_count,);
```
Comment (daniel-savu) [inline crates/issue/src/migration.rs:60]: nit
```suggestion
            log::info!(target: TARGET, "{} issue requests will be migrated.", issue_count,);
```

File: crates/issue/src/migration.rs (hunk 2 above)
Comment (daniel-savu) [reply crates/issue/src/migration.rs:12]: Ah it's because the v0 fields are annotated with macros

File: crates/issue/src/lib.rs [hunk 4]
```diff
@@ -348,6 +350,7 @@ impl<T: Config> Pallet<T> {
             amount: amount_user.amount(),
             fee: fee.amount(),
             griefing_collateral: griefing_collateral.amount(),
+            griefing_currency: griefing_collateral.currency(),
```
Comment (daniel-savu) [inline crates/issue/src/lib.rs:353]: Why not just change the type of `griefing_collateral` from `Balance` to `Amount`, to include `griefing_currency`?

Comment (daniel-savu) [reply crates/issue/src/lib.rs:353]: I'm assuming that it messes up storage byte alignment

File: crates/issue/src/migration.rs [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+    use super::*;
+    use crate::{
+        tests::{Test as T, *},
+        types::*,
+    };
+    use frame_support::bounded_vec;
+
+    #[allow(deprecated)]
+    #[test]
+    fn migration_works() {
```
Comment (daniel-savu) [inline crates/issue/src/migration.rs:128]: This does not test anything related to the issue request migration

File: standalone/runtime/tests/test_issue.rs [hunk 6]
```diff
@@ -667,101 +682,110 @@ mod execute_pending_issue_tests {
...
-                    vault.to_be_issued -= requested_btc;
-                })
-            );
-
-            // issue request is updated: status is complete
-            assert_eq!(
-                IssuePallet::issue_requests(issue_id).unwrap().status,
-                IssueRequestStatus::Completed
-            );
+            for griefing_currency in [DEFAULT_GRIEFING_CURRENCY, vault_id.collateral_currency()] {
```
Comment (daniel-savu) [inline standalone/runtime/tests/test_issue.rs:685]: According to the implementation, the griefing currency could be set to anything, not just vault collateral currencies. Not saying this test is wrong but it seems to assume that the griefing currency will not be e.g. `IBTC`

**Review by daniel-savu — APPROVED** [review]
My comments are not blocking

File: standalone/runtime/tests/test_issue.rs (hunk 6 above)
Comment (sander2) [reply standalone/runtime/tests/test_issue.rs:685]: It's main purpose was to test in the default griefing collateral plus one other one, it doesn't really matter which one it is. But maybe it makes more sense to use a hardcoded one

File: crates/issue/src/lib.rs (hunk 1 above)
Comment (sander2) [reply crates/issue/src/lib.rs:319]: redeem doesn't use it (except for an interaction with to-be-replaced tokens). Replace does use it, and we can change that in the future if we want, but not a prio for now

**Review by nud3l — APPROVED** [review]

**Pull request merged**